[exporter/fileexporter] Support for compressing the telemetry data before exporting (#14399)
open-telemetry/opentelemetry-collector-contrib: exporter/fileexporter/README.md

  - localtime : [default: false (use UTC)] whether or not the timestamps in backup files is formatted according to the host's local time.

- `format`[default: json]: define the data format of encoded telemetry data. The setting can be overridden with `proto`.

> - `compression`[default = none]: the compression algorithm used when exporting telemetry data to file. The options are: `none` and `zstd`

Reviewer: ```suggestion
- `compression`[no default]: the compression algorithm used when exporting telemetry data to file. Supported compression algorithms:`zstd`
```

  - localtime : [default: false (use UTC)] whether or not the timestamps in backup files is formatted according to the host's local time.

- `format`[default: json]: define the data format of encoded telemetry data. The setting can be overridden with `proto`.

> - `compression`[default:false]:defines whether to compress encoded telemetry data.

Reviewer: We may eventually want to support other compression algorithms. Instead of using a boolean, why don't we use a string? The default can be an empty string, and this will imply no compression. `"zstd"` can act the same as you currently have for `true`.

Author: Thanks for your advice～

For example, if your `path` is `data.json` and rotation is triggered, this file will be renamed to `data-2022-09-14T05-02-14.173.json`, and a new telemetry file created with `data.json`

## File Compression
Telemetry data is compressed according to the `compression` setting.
`fileexporter` does not compress data by default. 

Currently, `fileexporter` support the `zstd` compression algorithm, and we will support more compression algorithms in the future.

##  File Format 

Telemetry data is encoded according to the `format` setting and then written to the file.

When `format` is json and `compression` is none , telemetry data is written to file in JSON format. Each line in the file is a JSON object.

> Besides, each encoded object is preceded by 4 bytes (an unsigned 32 bit integer) which represent the number of bytes contained in the encoded object.When we need read the messages back in, we read the size, then read the bytes into a separate buffer, then parse from that buffer.

Reviewer: ```suggestion
Otherwise, when using `proto` format or any kind of encoding, each encoded object is preceded by 4 bytes (an unsigned 32 bit integer) which represent the number of bytes contained in the encoded object.When we need read the messages back in, we read the size, then read the bytes into a separate buffer, then parse from that buffer.
```

Author: Thanks
